## Java Exception 주의할 점

### 예외상황 (Exceptional Condition)

- 예외를 학습하면서 가장 빠지기 쉬운 오류는 **예외를 처리하는 방법**에 대해 집중한다는 것이다. 하지만 더 중요한 것은 **예외상황을 만들지 않는 코드를 작성하는 것**이다.
- Exception 이라고 하는 예외상황은 예측이 가능한 상황이다. 또한, 프로그램 실행중에 기본적으로는 일어나지 말아야할 상황이다. **프로그램을 하나의 함수집합이라고 생각을 했을때 함수 제 각각의 기능을 수행하는데 있어서 정상적으로 작동하지 못하게 되는 상황을 의미한다.** 즉, 정상적인 상황이라면 발생하지 상황이지만 일어날 가능성이 있는 상황이고 그것을 처리해줘야되는 상황인 것이다.

<br/>

### 예측하지 못한 오류상황

예외상황은 예측 가능한 상황이라고 하면 예측하지 못한 오류상황은 무엇인가. 그것은 버그라고 말할 수 있을 것이다. 예측하지 못했다면 그 상황에 대해 처리코드 또한 없을 것이고 프로그래머는 그 오류 상황에 대해 제대로 처리하는 코드를 작성하여 다시 빌드해야될 것이다. 즉, 원래 프로그램에 들어가야할 기능인데 없는 것이라고 볼 수 있다. 프로그램 실행도중에 발생하는 에러는 어쩔 수 없지만 예외는 프로그래머가 이에 대한 처리를 미리 해주어야 한다. 예외처리란 프로그램 실행시 발생할 수 있는 예기치 못한 예외의 발생에 대비한 코드를 작성하는 것이며 예외처리의 목적은 예외의 발생으로 인한 실행 중인 프로그램의 갑작스런 비정상 종료를 막고 정상적인 실행상태를 유지할 수 있도록 하는 것이다. 발생한 예외를 처리하지 못하면 프로그램은 비정상적으로 종료되며 처리되지 못한 예외는 JVM의 예외처리기가 받아서 예외의 원인을 화면에 출력한다. 

<br/>

### 예외는 진짜 예외 상황에만 사용하라

- 제어흐름용으로 예외를 사용해선 안된다

<br/>

### 복구할 수 있는 상황에는 체크 예외를, 프로그래밍 오류에는 런타임 예외를 사용하라

- 이는 체크 예외와 런타임 예외를 구분하는 가장 기본적인 규칙이다. 체크 예외를 던지면 호출하는 쪽이 그 예외를 catch로 잡아 처리하거나 외부로 전파하도록 강제하게 된다. API 설계자는 API 사용자에게 체크 예외를 던져주어 그 상황에서 회복해내라고 요구한 것이다. 물론 사용자는 예외를 catch만 하고 별다른 조치를 취하지 않을 수 있지만 이는 보통 좋지 않은 생각이다. 

<br/>

### e.**printStackTrace**()를 사용하지 마라

- 개발시에만 사용할 것, 불필요한 로그 출력으로 메모리 누수 발생

<br/>

## JVM은 Exception을 어떻게 처리하는가
[참고: Geeksforgeeks Exception in Java](https://www.geeksforgeeks.org/exceptions-in-java/)

### 기본 예외 처리

메소드 내부에서 익셉션이 발생할 때마다 메소드는 `Exception Object`라고 하는 객체를 생성하여 JVM에 전달한다. `Exception Object`는 익셉션의 이름과 설명, 익셉션이 발생한 프로그램의 현재 상태를 포함한다. `Exception Object`를 만들고 JVM에서 처리하는 것을 Exception throw라고 한다. 익셉션이 발생한 메소드에 도달하기 위해 호출된 메소드 목록이 여러개 있을 수 있으며 이렇게 정렬된 메소드 목록을 `Call stack`이라고 합니다. 이제 익셉션이 발생한 때에 상황을 살펴보자. 

- JVM은 호출 스택을 검색하여 발생한 익셉션을 처리할 수 있는 코드 블록이 포함된 메서드를 찾는다. 이 코드 블록을 예외 핸들러라고 한다.
- JVM은 익셉션이 발생한 메소드부터 검색을 시작하고 메소드가 호출된 역순으로 call stack을 진행한다.
- 적절한 핸들러를 찾으면 발생한 익셉션을 핸들러에 전달한다. 적절한 핸들러는 던져진 `Exception Object`의 타입이 처리할 수 있는 타입과 일치한다는 것을 의미합니다.
- JVM이 `Call stack`의 모든 메소드를 검색하고 적절한 핸들러를 찾지 못한 경우 JVM은 JVM의 일부인 기본 익셉션 핸들러에 `Exception Object`를 넘겨준다. 이 핸들러는 예외 정보를 다음과 같은 형식으로 출력하고 프로그램을 비정상 종료한다.

```java
Exception in thread "xxx" Name of Exception : Description
... ...... ..  // Call Stack
```